examples/src/scala/org/pantsbuild/example/strict_deps_jvm.md

```diff
@@ -0,0 +1,228 @@
+Strict Dependencies
+===================
+
+### What is strict_deps?
+
+Historically, Pants works with transitive dependencies. This means that in
+order to compile each target, the compiler must scan that full transitive set.
+But, most targets do not need access to all of their transitive dependencies at
+compile time. We can make compilation faster if we can reduce those compile
+dependencies to just the necessary ones.
+
+Strict_deps is a Pants feature that changes compile behavior so that only
+direct dependencies and exports are included on the compile classpath for a
+target. This reduces scalac times by reducing the amount of class files scalac
+needs to scan before beginning a compile. It improves cache hit rates by
+reducing the number of files that are included in each target’s compile cache
+key.eat
+
+### Why strict_deps?
+
+Use of strict_deps improves cache hit rates and per-target compilation time
+(on the order of 10-20%).
+
+Example
+-------
+
+With strict_deps enabled, we can identified some gaps between what the
+compiler expects on a classpath and what is actually available on the
+classpath as specified in BUILD files.
+
+Let's look at a simple example involving 3 targets A, B and C. B depends on A,
+and C depends on B. Thus their dependency graph can be represented like this:
+C->B->A. Their declarations in BUILD file are shown as below:
+
+    :::python
+    scala_library(
+        name='A',
+        source=['A.scala'],
+    )
+
+    scala_library(
+        name='B',
+        source=['B.scala'],
+        dependencies=[':A'],
+    )
+
+    scala_library(
+        name='C',
+        source=['C.scala'],
+        dependencies=[':B'],
+    )
+
+If we compile C with strict_deps turned off, everything works fine, as all
+transitive dependencies of C will be added to compile time classpaths.
+However, the compilation may require additional dependencies if strict_deps
+is turned on. With strict_deps=True, only B is in C's compile time classpath.
+There are 2 cases for the failure.
+
+1. C directly uses APIs from A. This is obviously a mistake by author of C,
+and the fix is simple, just add A to C's dependency list.
+
+2. C does not use APIs from A, but compiler still wants A to be in C's
+classpaths. This is tricky, as author of C may not even be aware of A. It's
+hard for C's author to get the right dependency list in this case.
+
+People may wonder why compiler asks for A to compile C in the first place.
+Unfortunately this is a limitation for both javac and scalac. This is just
+how these compilers behave.
```

Comment:
Sorry, I meant to include the "For instance" - this is just one example of why a transitive dep may be needed, there are others too :) Maybe this could be phrased:

> People may wonder why the compiler asks for A to compile C in the first place. There are a number of reasons for this, an example of which is that the java compiler needs all transitively implemented interfaces to be on the classpath so that it can resolve default method implementations.